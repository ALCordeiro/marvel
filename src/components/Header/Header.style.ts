import styled from 'styled-components';

export const HeaderContainer = styled.div`
    display: flex;
    justify-content: center;
    flex-direction: column;
    align-items: center;
    padding-bottom: 40px;
`;

export const LogoContainer = styled.div`
    display: flex;
    max-width: 330px;
    max-height: 120px;
    justify-content: center;
    padding-bottom: 20px;
`;

export const TitleHeader = styled.h2`
    text-transform: uppercase;
    font-size: 38px;
    color: #202020;
    padding-bottom: 10px;
`;

export const SubtitleHeader = styled.h3`
    font-size: 16px;
    color: gray;
    font-weight: 400;
    padding-bottom: 50px;
`;

export const SearchBoxContainer = styled.div`
    position: relative;
    width: 75%;
    height: 50px;
`;

export const SearchBox = styled.input.attrs({
    type: 'text',
    placeholder: 'Procure por herÃ³is',
    name: 'search',
    id: 'search-input-characters',
})`
    width: 100%;
    border-radius: 15px;
    border: 0px;
    background-color: rgba(255, 0, 0, 0.1);
    padding: 5px 5px 5px 45px;
    position: absolute;
    top: 0;
    left: 0;
    z-index: 5;
    height: 100%;
    border-bottom-left-radius: 50px;
    border-bottom-right-radius: 50px;
    border-top-left-radius: 50px;
    border-top-right-radius: 50px;
`;

interface SearchIcon {
    searchIcon: string;
}

const urlImg = (props: SearchIcon) => `${props.searchIcon}`;

export const SearchBoxCorner = styled.input.attrs({
    type: 'button',
    value: '',
})`
    position: absolute;
    top: 14px;
    left: 14px;
    height: 20px;
    width: 20px;
    z-index: 10;
    border-radius: 10px;
    border: none;
    background: no-repeat center url(${urlImg});
    cursor: pointer;
`;
