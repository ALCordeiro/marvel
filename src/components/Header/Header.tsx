/* eslint-disable function-paren-newline */
import React, { useState } from 'react';

import { useDispatch } from 'react-redux';
import {
    HeaderContainer,
    LogoContainer,
    TitleHeader,
    SubtitleHeader,
    SearchBox,
    SearchBoxContainer,
    SearchBoxCorner,
} from './Header.style';
import {
    ICharacters2,
    ICharacterMap,
} from '../../interfaces/characters.interface';
import { searchCharactersRequest } from '../../store/ducks/characters/actions';

export const HeaderComponent: React.FC<ICharacters2> = ({
    characters,
}: ICharacters2) => {
    const [search, setSearch] = useState('');
    const dispatch = useDispatch();

    function handleSearch() {
        const arrFiltered = characters.filter((el: ICharacterMap) =>
            el.name.toLowerCase().includes(search.toLowerCase()),
        );
        dispatch(searchCharactersRequest(arrFiltered, search));
    }

    return (
        <HeaderContainer>
            <LogoContainer>
                <img
                    src={`${window.location.origin}/assets/marvel-logo.png`}
                    alt="marvel-logo"
                />
            </LogoContainer>
            <TitleHeader>Explore o Universo</TitleHeader>
            <SubtitleHeader>
                Mergulhe no domínio deslumbrante de todos os personagens que
                você ama - e aqueles que você descobrirá em breve!
            </SubtitleHeader>

            <SearchBoxContainer>
                <SearchBox onChange={e => setSearch(e.target.value)} />
                <SearchBoxCorner
                    searchIcon={`${window.location.origin}/assets/search.svg`}
                    onClick={() => handleSearch()}
                />
            </SearchBoxContainer>
        </HeaderContainer>
    );
};

export default HeaderComponent;
