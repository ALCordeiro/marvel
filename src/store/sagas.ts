import { all } from 'redux-saga/effects';
import { CharactersSaga } from './ducks/characters/sagas';

export function* rootSagas() {
    yield all([...CharactersSaga]);
}
