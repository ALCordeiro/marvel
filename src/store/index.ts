import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { CharactersState } from './ducks/characters/types';
import rootReducer from './ducks/rootReducer';
import { rootSagas } from './sagas';

export interface ApplicationState {
    characters: CharactersState;
}

const middleware = [];
const initialState = {};
const sagaMiddleware = createSagaMiddleware();
middleware.push(sagaMiddleware);

const store = createStore(
    rootReducer,
    initialState,
    composeWithDevTools(applyMiddleware(...middleware)),
);

sagaMiddleware.run(rootSagas);

export default store;
