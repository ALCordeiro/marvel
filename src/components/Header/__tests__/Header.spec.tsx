/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { HeaderComponent } from '../Header';
import store from '../../../store';

const characters: any = [
    {
        comics: {
            available: 12,
            collectionURI:
                'http://gateway.marvel.com/v1/public/characters/1011334/comics',
            items: Array(12),
            returned: 12,
        },
        description: '',
        events: {
            available: 1,
            collectionURI:
                'http://gateway.marvel.com/v1/public/characters/1011334/events',
            items: Array(1),
            returned: 1,
        },
        id: 1011334,
        modified: '2014-04-29T14:18:17-0400',
        name: '3-D Man',
        resourceURI: 'http://gateway.marvel.com/v1/public/characters/1011334',
        series: {
            available: 3,
            collectionURI:
                'http://gateway.marvel.com/v1/public/characters/1011334/series',
            items: Array(3),
            returned: 3,
        },
        stories: {
            available: 21,
            collectionURI:
                'http://gateway.marvel.com/v1/public/characters/1011334/stories',
            items: Array(20),
            returned: 20,
        },
        thumbnail: {
            path: 'http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784',
            extension: 'jpg',
        },
        map: () => 'test',
        length: () => 1,
        filter: () => 'test',
    },
];

let wrapper;

beforeEach(() => {
    wrapper = render(
        <Provider store={store}>
            <Router>
                <HeaderComponent characters={characters} />
            </Router>
        </Provider>,
    );
});

test('should render', async () => {
    const { container } = wrapper;
    expect(container.firstChild);
});

test('should match snapshot', () => {
    const { container } = wrapper;
    expect(container.firstChild).toMatchSnapshot();
});

test('should render with a div as first child', () => {
    const { container } = wrapper;
    expect(container.firstChild.nodeName).toBe('DIV');
});

test('should show the correct text for h2', () => {
    expect(screen.getByRole('heading', { level: 2 }).textContent).toBe(
        'Explore o Universo',
    );
});

test('should show the correct text for h3', () => {
    expect(screen.getByRole('heading', { level: 3 }).textContent).toBe(
        'Mergulhe no domínio deslumbrante de todos os personagens que você ama - e aqueles que você descobrirá em breve!',
    );
});
