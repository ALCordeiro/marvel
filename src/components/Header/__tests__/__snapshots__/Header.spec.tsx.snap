// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  class="sc-bdnxRM fGhsRl"
>
  <div
    class="sc-gtsrHT bsaVBs"
  >
    <img
      alt="marvel-logo"
      src="http://localhost/assets/marvel-logo.png"
    />
  </div>
  <h2
    class="sc-dlnjwi cDcEdw"
  >
    Explore o Universo
  </h2>
  <h3
    class="sc-hKFxyN kAroCE"
  >
    Mergulhe no domínio deslumbrante de todos os personagens que você ama - e aqueles que você descobrirá em breve!
  </h3>
  <div
    class="sc-eCApnc cCPDbx"
  >
    <input
      class="sc-jSFjdj eZWXu"
      id="search-input-characters"
      name="search"
      placeholder="Procure por heróis"
      type="text"
    />
    <input
      class="sc-gKAaRy bwuAgQ"
      type="button"
      value=""
    />
  </div>
</div>
`;
